monkey/monkey_island/cc/ui/src/components/report-components/ransomware/BreachSection.tsx

```diff
@@ -19,21 +19,49 @@ const BREACH_DESCRIPTION = <>
                            </>
 
 function BreachSection() {
+  const [agents, setAgents] = useState(null);
   const [machines, setMachines] = useState(null);
 
   useEffect(() => {
-      IslandHttpClient.get(APIEndpoint.manual_exploitation)
-      .then(resp => setMachines(resp.body['manual_exploitations']));
+    getAllAgents().then(agents => setAgents(agents));
+    getAllMachines().then(machines => setMachines(machines));
   }, []);
 
-  if(machines !== null){
-    let body = getBreachSectionBody(machines);
+  if((machines !== null) && (agents !== null)){
+    let manuallyExploitedMachines = getManuallyExploitedMachines(agents, machines);
+    let body = getBreachSectionBody(manuallyExploitedMachines);
     return (<NumberedReportSection index={1} title={'Breach'} description={BREACH_DESCRIPTION} body={body}/>)
   } else {
     return <LoadingIcon />
   }
 }
 
+function getManuallyExploitedMachines(agents, machines){
+  let manuallyExploitedMachines = [];
+  let manuallyStartedAgents = getManuallyStartedAgents(agents);
+  for (let agent of manuallyStartedAgents) {
+    let machine = getMachineByAgent(agent, machines);
+    if (machine !== null){
+      let manualExploitatedMachine = {};
+      manualExploitatedMachine['hostname'] = getMachineHostname(machine);
+      manualExploitatedMachine['ip_addresses'] = machine['network_interfaces'].map(network_interface => network_interface.split('/')[0]);
```

Comment:
We're duplicating This logic all over the place. See also #2551. Let's make his reusable.